task_sdk/src/airflow/sdk/execution_time/context.py

```diff
@@ -281,6 +283,84 @@ def _get_asset_from_db(name: str | None = None, uri: str | None = None) -> Asset
         return Asset(**msg.model_dump(exclude={"type"}))
 
 
+@attrs.define(init=False)
+class InletEventsAccessors(Mapping[Union[int, Asset, AssetAlias, AssetRef], Any]):
+    _inlets: list[Any]
+    _assets: dict[AssetUniqueKey, Asset]
+    _asset_aliases: dict[AssetAliasUniqueKey, AssetAlias]
+
+    def __init__(self, inlets: list) -> None:
+        self._inlets = inlets
+        self._assets = {}
+        self._asset_aliases = {}
+
+        for inlet in inlets:
+            if isinstance(inlet, Asset):
+                self._assets[AssetUniqueKey.from_asset(inlet)] = inlet
+            elif isinstance(inlet, AssetAlias):
+                self._asset_aliases[AssetAliasUniqueKey.from_asset_alias(inlet)] = inlet
+            elif isinstance(inlet, AssetNameRef):
+                asset = OutletEventAccessors._get_asset_from_db(name=inlet.name)
+                self._assets[AssetUniqueKey.from_asset(asset)] = asset
+            elif isinstance(inlet, AssetUriRef):
+                asset = OutletEventAccessors._get_asset_from_db(uri=inlet.uri)
+                self._assets[AssetUniqueKey.from_asset(asset)] = asset
+
+    def __iter__(self) -> Iterator[Asset | AssetAlias]:
+        return iter(self._inlets)
+
+    def __len__(self) -> int:
+        return len(self._inlets)
+
+    def __getitem__(self, key: int | Asset | AssetAlias | AssetRef):
+        from airflow.sdk.definitions.asset import Asset
+
+        if isinstance(key, int):  # Support index access; it's easier for trivial cases.
+            obj = self._inlets[key]
+            if not isinstance(obj, (Asset, AssetAlias, AssetRef)):
+                raise IndexError(key)
+        else:
+            obj = key
+
+        return self._get_asset_events_from_db(obj)
+
+    # TODO: This is temporary to avoid code duplication between here & airflow/models/taskinstance.py
```

Comment:
Should probably explain what is to do next

Reply:
I think this relates to https://github.com/apache/airflow/pull/45960/files#r1967152850. This follows https://github.com/apache/airflow/pull/45727/files#diff-0274e1490fa6b4fc17f3d1e06a5037d2d15d699322c554ff901aefa013bc6e89R260. 

I guess we still need to do something in task sdk to make `from airflow.sdk.execution_time.task_runner import SUPERVISOR_COMMS` importable in tests or some other places. But would like to confirm with @ashb and @amoghrajesh whether my understanding is correct.

Reply:
I don't quite understand the question. Can you elaborate on it a bit please?

Also we shouldn't really call this method "_from_db" -- as its not hitting the DB, it's hitting the API server

Reply:
This code isn't duplicated between models/ti and this though, not really?

Reply:
I thought we wanted to keep this logic in task_sdk at the end and remove the one in `airflow.utils.context` (same for outlet). 

I tried to do it this way but encountered an import error when tests reach https://github.com/apache/airflow/pull/45960/files#diff-0274e1490fa6b4fc17f3d1e06a5037d2d15d699322c554ff901aefa013bc6e89R334

Reply:
so my guess of why we did it is because we still cannot do `from airflow.sdk.execution_time.task_runner import SUPERVISOR_COMMS` because of other task SDK issus

Reply:
This also fails on the outlet end as well

Reply:
It's now removed